lib/WorkDoneProgress/MessageProgressNotifier.php

```diff
@@ -16,45 +12,62 @@ final class MessageProgressNotifier implements ProgressNotifier
      */
     private $api;
 
-    public function __construct(ClientApi $api)
-    {
-        $this->api = $api->window()->showMessage();
-    }
-
     /**
-     * {@inheritDoc}
+     * @var string The title of the progress being reported
      */
-    public function create(WorkDoneToken $token): Promise
+    private $title;
+
+    public function __construct(ClientApi $api)
     {
-        return new Success(new ResponseMessage(
-            Uuid::uuid4(),
-            null,
-        ));
+        $this->api = $api->window()->showMessage();
     }
 
     /**
      * {@inheritDoc}
      */
     public function begin(
-        WorkDoneToken $token,
         string $title,
         ?string $message = null,
         ?int $percentage = null,
         ?bool $cancellable = null
     ): void {
-        $this->api->info($message);
+        $this->title = $title;
```

Comment:
not sure this is a good idea, does it imply a temporal coupling? or you could write a new `report` but use a previous title.

an alternative would be to return an object `begin(...): ProgressTracker`, where progress tracker contains the necessary dependencies.

but I actually didn't mind the first approach, what was wrong with it?

Reply:
For me it's not a temporal coupling, it's a logic one.
The progress is a all, from the begin to the end request.
In a classic `$/progress` it's the client which does the job of matching each subsequent requests to the original `begin` one through the token.
Since here we are attempting to have a fallback strategy with a method with does not have this link between subsequent request we have to deal with it ourselves.

On the next instance of notifier, or simply the next `begin` the title will be updated.
At start I had implemented a  `WorkDoneProgress` class which was holding these information and a tracker which linked each object to it's token so that we can even ensure `begin` was only called once, that the progress was destroyed on `end` but it was a bit complicated back then for a first draft.

With the current implementation we should use a different notifier instances for different reports.
So there is no risk of the same title being used for different reports.

For me those are implementations details hidden behind the factory, so there is no reason to use them manually.
I should mark them as internal then I guess to make it clearer.
If we want to enforce the fact that a notifier can be use for only one report we would need to add additional logic.
I was afraid it would add too much complexity just to prevent an invalid use case, so I tried to add it behind abstraction layers instead.

I'm not against reworking the notifiers to make them bullet proof to mistakes if you want, it's quite interesting actually :)